File: 2021/8/main.hs
ex1 :: [[String]] -> [[String]] -> Int
ex1 _ output = sum [ sum [ if elem (length digit) [2,3,4,7] then 1 else 0| digit <- line] |line <- output] --can rewrite as filter

intersection :: String -> String -> Int
intersection word1 word2 = sum [if elem letter word2 then 1 else 0 | letter <- word1] --like above

getUnique :: [(Int,String)] -> Int -> String
getUnique lengths len = last . head . filter (\(l, val) -> l == len) $ lengths
    
findDigit :: String -> [(Int,String)] -> String
findDigit number lengths = case length number of
    2 -> "1"
    4 -> "4"
    3 -> "7"
    7 -> "8"
    6 -> if intersection number (getUnique lengths 4) == 4 then "9" else if intersection number (getUnique lengths 2) == 1 then "6" else "0"
    5 -> if intersection number (getUnique lengths 4) == 2 then "2" else if intersection number (getUnique lengths 2) == 2 then "3" else "5"

findNumber :: [String] -> [String] -> Int
findNumber digits output = read (findDigit (head digits) lengths ++ findDigit (digits!!1) lengths ++ findDigit (digits!!2) lengths ++ findDigit (last digits) lengths) :: Int
    where lengths = [(length digit, digit) | digit <- digits]

engine :: [[String]] -> [[String]] -> Int -> Int
engine []             []            sum = sum
engine (digit:digits) (line:output) sum = engine digits output (sum+newSum)
    where newSum = findNumber digit line

ex2 :: [[String]] -> [[String]] -> Int
ex2 digits output = engine digits output 0

parse :: [String] -> ([[String]], [[String]])
parse input = ([words . head $ line | line <- divided] , [words . last $ line | line <- divided])
    where divided = [wordsWhen (=='|') line | line <- input]

wordsWhen :: (Char -> Bool) -> String -> [String]
wordsWhen p s = case dropWhile p s of
    "" -> []
    s' -> w : wordsWhen p s''
        where (w, s'') = break p s'

main :: IO ()
main = do
    strLines <- readFile "./8/data.txt"
    let (digits, output) = parse . lines $ strLines
    print $ ex1 digits output
    print $ ex2 digits output

File: 2021/8/main.py
def ex1(digits, output): #digit k uses (->) n signals: 1 -> 2, 4 -> 4, 7 -> 3, 8 -> 7
    return sum( [ sum( [ 1 if len(digit) in (2,4,3,7) else 0 for digit in line] ) for line in output ] )

def ex2(digits, output):
    res = 0
    for i in range(0, len(digits)):
        lengths = {len(digit): set(digit) for digit in digits[i]} # here we only care about unique (with unique length) numbers to help us distinguish between those not unique 
        n = ''
        for number in output[i]:
            num = set(number)
            if len(num) == 2:
                n += '1'
            elif len(num) == 4:
                n += '4'
            elif len(num) == 3:
                n += '7'
            elif len(num) == 7:
                n += '8'
            elif len(num) == 6:
                if len(num.intersection(lengths.get(4))) == 4: # 4 -> 4
                    n += '9'
                elif len(num.intersection(lengths.get(2))) == 1: # 2 -> 1
                    n += '6'
                else:
                    n += '0'
            else:
                if len(num.intersection(lengths.get(4))) == 2: # 2 -> 1
                    n += '2'
                elif len(num.intersection(lengths.get(2))) == 2: # 4 -> 4
                    n += '3'
                else:
                    n += '5'
        res += int(n)
    return res
        
with open("./8/data.txt", "r") as f:
    lines = [row.split("|") for row in f]
    digits = [line[0].split() for line in lines]
    output = [line[1].split() for line in lines]

print(ex1(digits, output))
print(ex2(digits, output))
'''
digit k uses n signals [a,b,c,d,e,f,g]
0: 6
1: 2 unique
2: 5
3: 5
4: 4 unique
5: 5
6: 6
7: 3 unique
8: 7 unique
9: 6

     069
06         09
      69
069       069
     069

     235
5           23
     235
2           35
     235
'''